### PR TITLE
add kernel in functions all, any

### PR DESCRIPTION
**Pull request by Rubtsowa**
tests/test_logic.py::test_all[(0,)-float64] 
tests/test_logic.py::test_all[(0,)-float32] 
tests/test_logic.py::test_all[(0,)-int64] 
tests/test_logic.py::test_all[(0,)-int32] 
tests/test_logic.py::test_all[(0,)-bool] 
tests/test_logic.py::test_all[(0,)-bool_] 
tests/test_logic.py::test_all[(4,)-float64] 
tests/test_logic.py::test_all[(4,)-float32] 
tests/test_logic.py::test_all[(4,)-int64] 
tests/test_logic.py::test_all[(4,)-int32] 
tests/test_logic.py::test_all[(4,)-bool] 
tests/test_logic.py::test_all[(4,)-bool_] 
tests/test_logic.py::test_all[(2,3)-float64] 
tests/test_logic.py::test_all[(2,3)-float32] 
tests/test_logic.py::test_all[(2,3)-int64] 
tests/test_logic.py::test_all[(2,3)-int32] 
tests/test_logic.py::test_all[(2,3)-bool] 
tests/test_logic.py::test_all[(2,3)-bool_] 
tests/test_logic.py::test_all[(2,2,2)-float64] 
tests/test_logic.py::test_all[(2,2,2)-float32] 
tests/test_logic.py::test_all[(2,2,2)-int64] 
tests/test_logic.py::test_all[(2,2,2)-int32] 
tests/test_logic.py::test_all[(2,2,2)-bool] 
tests/test_logic.py::test_all[(2,2,2)-bool_] 
tests/test_logic.py::test_any[(4,)-float64] 
tests/test_logic.py::test_any[(4,)-float32] 
tests/test_logic.py::test_any[(4,)-int64] 
tests/test_logic.py::test_any[(4,)-int32] 
tests/test_logic.py::test_any[(4,)-bool] 
tests/test_logic.py::test_any[(4,)-bool_] 
tests/test_logic.py::test_any[(2,3)-float64] 
tests/test_logic.py::test_any[(2,3)-float32] 
tests/test_logic.py::test_any[(2,3)-int64] 
tests/test_logic.py::test_any[(2,3)-int32] 
tests/test_logic.py::test_any[(2,3)-bool] 
tests/test_logic.py::test_any[(2,3)-bool_] 
tests/test_logic.py::test_any[(2,2,2)-float64] 
tests/test_logic.py::test_any[(2,2,2)-float32] 
tests/test_logic.py::test_any[(2,2,2)-int64] 
tests/test_logic.py::test_any[(2,2,2)-int32] 
tests/test_logic.py::test_any[(2,2,2)-bool] 
tests/test_logic.py::test_any[(2,2,2)-bool_]

### CODE REVIEW

File: dpnp/backend/include/dpnp_iface.hpp [hunk 1]
```diff
@@ -113,7 +113,7 @@ void dpnp_memory_memcpy_c(void* dst, const void* src, size_t size_in_bytes);
  * @param [out] result      Output array.
  * @param [in]  size        Number of input elements in `array`.
  */
-template <typename _DataType, typename _ResultType>
+template <typename _DataType>
```
Comment (samir-nasibli) [inline dpnp/backend/include/dpnp_iface.hpp:116]: Leave it as it was. See, https://github.com/IntelPython/dpnp/pull/637#discussion_r586731916
The comment was for `any` but it suits also for `all`.

File: dpnp/backend/include/dpnp_iface.hpp [hunk 2]
```diff
@@ -124,7 +124,7 @@ INP_DLLEXPORT void dpnp_all_c(void* array, void* result, const size_t size);
...
```
Comment (samir-nasibli) [inline dpnp/backend/include/dpnp_iface.hpp:127]: Leave it as it was. See, https://github.com/IntelPython/dpnp/pull/637#discussion_r586731916

File: dpnp/backend/kernels/dpnp_krnl_logic.cpp [hunk 3]
```diff
@@ -30,68 +30,89 @@
...
+    cl::sycl::event event;
+
+    const _DataType* array_in = reinterpret_cast<const _DataType*>(array1_in);
+    bool* result = reinterpret_cast<bool*>(result1);
 
-    bool res = true;
+    result[0] = true;
 
-    for (size_t i = 0; i < size; ++i)
+    if ((array1_in != nullptr) && (result1 != nullptr))
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_logic.cpp:46]: Do check before, on top.

File: dpnp/backend/kernels/dpnp_krnl_logic.cpp [hunk 4]
```diff
@@ -30,68 +30,89 @@
...
 
     const _DataType* array_in = reinterpret_cast<const _DataType*>(array1_in);
-    _ResultType* result = reinterpret_cast<_ResultType*>(result1);
+    bool* result = reinterpret_cast<bool*>(result1);
 
-    bool res = false;
+    result[0] = false;
 
-    for (size_t i = 0; i < size; ++i)
+    if ((array1_in != nullptr) && (result1 != nullptr))
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_logic.cpp:81]: What if `result1` is nullptr?
You initilizing before check - it is bad.

move nullptr check on top.

**Review by samir-nasibli — COMMENTED** [review]
numpy tests!

File: dpnp/backend/kernels/dpnp_krnl_logic.cpp [hunk 5]
```diff
@@ -30,68 +30,89 @@
...
-        }
-    }
+        cl::sycl::range<1> gws(size);
+        auto kernel_parallel_for_func = [=](cl::sycl::id<1> global_id) {
+            size_t i = global_id[0];
 
-    result[0] = res;
+            if (!array_in[i])
+            {
+                result[0] = false;
```
Comment (shssf) [inline dpnp/backend/kernels/dpnp_krnl_logic.cpp:54]: Sorry but I think it will not work in parallel environment.
The problem is in accessing the same element from all threads. 
Let's imagine input array with 100 elements. This code emits 100 threads (gws=100). All threads will try to write their value `true` or `false` into the same memory address (result[0]).
Final result will strongly depend on the order of threads execution. Strictly speaking, the result will be overwritten by last active thread.

Are you sure this algorithm should return only one value for all number of input values? I didn't understand this by reading numpy documentation. They mentioned scalar and array as output at the same time.

Comment (shssf) [reply dpnp/backend/kernels/dpnp_krnl_logic.cpp:54]: No, I'm wrong. It will work. Dismiss my previous comment.

File: dpnp/backend/kernels/dpnp_krnl_logic.cpp (hunk 4 above)
Comment (shssf) [reply dpnp/backend/kernels/dpnp_krnl_logic.cpp:81]: "What if result1 is nullptr?"
Behavior unpredictable I think. The code will wait on uninitialized `event` by `event.wait()`

Comment (shssf) [inline dpnp/backend/kernels/dpnp_krnl_logic.cpp:81]:
```suggestion
    if ((array1_in == nullptr) || (result1 == nullptr))
    {
        return;
    }
```

File: dpnp/backend/kernels/dpnp_krnl_logic.cpp [hunk 6]
```diff
@@ -29,53 +29,79 @@
...
+
+    cl::sycl::range<1> gws(size);
+    auto kernel_parallel_for_func = [=](cl::sycl::id<1> global_id) {
+        size_t i = global_id[0];
+
         if (array_in[i])
         {
-            res = true;
-            break;
+            result[0] = true;
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_logic.cpp:94]: I am not sure but will this work correct?
In any way bad design for parallel work.
You may need big array (size >200 for example) for the check.

@shssf why do you think it is OK? https://github.com/IntelPython/dpnp/pull/640#discussion_r588896344

File: dpnp/backend/kernels/dpnp_krnl_logic.cpp [hunk 7]
```diff
@@ -29,53 +29,79 @@
...
     _ResultType* result = reinterpret_cast<_ResultType*>(result1);
 
-    bool res = true;
+    result[0] = true;
 
-    for (size_t i = 0; i < size; ++i)
+    if ((array1_in == nullptr) || (result1 == nullptr))
     {
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (Rubtsowa) [reply dpnp/backend/kernels/dpnp_krnl_logic.cpp:49]: Sorry, but in this case numpy returning True, not empty array. 

>>> a=np.array([])
>>> a.all()
True

File: dpnp/backend/kernels/dpnp_krnl_logic.cpp [hunk 8]
```diff
@@ -29,53 +29,79 @@
...
     _ResultType* result = reinterpret_cast<_ResultType*>(result1);
 
-    bool res = false;
+    result[0] = false;
 
-    for (size_t i = 0; i < size; ++i)
+    if ((array1_in == nullptr) || (result1 == nullptr))
     {
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (samir-nasibli) [reply dpnp/backend/kernels/dpnp_krnl_logic.cpp:86]: So what?
You never test this case from python level.
From python (cython) level you won't call a function (`dpnp_any_c`) with null ptrs due to your design.

Proposed changes is needed for dpnpc native users.

File: dpnp/backend/kernels/dpnp_krnl_logic.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (samir-nasibli) [reply dpnp/backend/kernels/dpnp_krnl_logic.cpp:49]: see https://github.com/IntelPython/dpnp/pull/640#discussion_r591419452

File: dpnp/backend/kernels/dpnp_krnl_logic.cpp (hunk 6 above)
Comment (samir-nasibli) [reply dpnp/backend/kernels/dpnp_krnl_logic.cpp:94]: I get it too:) it will work.

File: dpnp/backend/kernels/dpnp_krnl_logic.cpp [hunk 9]
```diff
@@ -29,53 +29,79 @@
...
+        size_t i = global_id[0];
+
         if (!array_in[i])
         {
-            res = false;
-            break;
+            result[0] = false;
         }
-    }
+    };
```
Comment (shssf) [inline dpnp/backend/kernels/dpnp_krnl_logic.cpp:59]:
```suggestion
    }
```

**Review by shssf — APPROVED** [review]

File: dpnp/backend/kernels/dpnp_krnl_logic.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (shssf) [reply dpnp/backend/kernels/dpnp_krnl_logic.cpp:86]: this is related to backend testing. Perhaps it is good idea to start writing some backend tests in `gtest` suite

**Review by shssf — APPROVED** [review]

File: dpnp/backend/kernels/dpnp_krnl_logic.cpp [hunk 10]
```diff
@@ -29,53 +29,80 @@
 #include "dpnp_iface.hpp"
 #include "queue_sycl.hpp"
 
+
+template <typename _DataType, typename _ResultType>
+class dpnp_all_c_kernel;
+
 template <typename _DataType, typename _ResultType>
 void dpnp_all_c(void* array1_in, void* result1, const size_t size)
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_logic.cpp:37]:
```suggestion
void dpnp_all_c(const void* array1_in, void* result1, const size_t size)
```
Do the same in header file

**Comment by samir-nasibli** [comment]
> tests/test_logic.py::test_all[(4,)-float64]
> tests/test_logic.py::test_all[(4,)-float32]
> tests/test_logic.py::test_all[(4,)-int64]
> tests/test_logic.py::test_all[(4,)-int32]
> tests/test_logic.py::test_all[(4,)-bool]
> tests/test_logic.py::test_all[(4,)-bool_]
> tests/test_logic.py::test_all[(2,3)-float64]
> tests/test_logic.py::test_all[(2,3)-float32]
> tests/test_logic.py::test_all[(2,3)-int64]
> tests/test_logic.py::test_all[(2,3)-int32]
> tests/test_logic.py::test_all[(2,3)-bool]
> tests/test_logic.py::test_all[(2,3)-bool_]
> tests/test_logic.py::test_all[(2,2,2)-float64]
> tests/test_logic.py::test_all[(2,2,2)-float32]
> tests/test_logic.py::test_all[(2,2,2)-int64]
> tests/test_logic.py::test_all[(2,2,2)-int32]
> tests/test_logic.py::test_all[(2,2,2)-bool]
> tests/test_logic.py::test_all[(2,2,2)-bool_]
> tests/test_logic.py::test_any[(4,)-float64]
> tests/test_logic.py::test_any[(4,)-float32]
> tests/test_logic.py::test_any[(4,)-int64]
> tests/test_logic.py::test_any[(4,)-int32]
> tests/test_logic.py::test_any[(4,)-bool]
> tests/test_logic.py::test_any[(4,)-bool_]
> tests/test_logic.py::test_any[(2,3)-float64]
> tests/test_logic.py::test_any[(2,3)-float32]
> tests/test_logic.py::test_any[(2,3)-int64]
> tests/test_logic.py::test_any[(2,3)-int32]
> tests/test_logic.py::test_any[(2,3)-bool]
> tests/test_logic.py::test_any[(2,3)-bool_]
> tests/test_logic.py::test_any[(2,2,2)-float64]
> tests/test_logic.py::test_any[(2,2,2)-float32]
> tests/test_logic.py::test_any[(2,2,2)-int64]
> tests/test_logic.py::test_any[(2,2,2)-int32]
> tests/test_logic.py::test_any[(2,2,2)-bool]
> tests/test_logic.py::test_any[(2,2,2)-bool_]

add, please also tests for empty arrays.

File: tests/test_logic.py [hunk 11]
```diff
@@ -42,8 +42,8 @@ def test_all(type, shape):
                          [numpy.float64, numpy.float32, numpy.int64, numpy.int32, numpy.bool, numpy.bool_],
                          ids=['float64', 'float32', 'int64', 'int32', 'bool', 'bool_'])
 @pytest.mark.parametrize("shape",
-                         [(4,), (2, 3), (2, 2, 2)],
-                         ids=['(4,)', '(2,3)', '(2,2,2)'])
+                         [(0,), (4,), (2, 3), (2, 2, 2)],
+                         ids=['(0,)', '(4,)', '(2,3)', '(2,2,2)'])
 def test_any(type, shape):
     size = 1
     for i in range(len(shape)):
```
Comment (samir-nasibli) [inline tests/test_logic.py:49]: `dparray` or `ndarray` have their own method, that returns array data size.

File: tests/test_logic.py [hunk 12]
```diff
@@ -42,8 +42,8 @@ def test_all(type, shape):
                          [numpy.float64, numpy.float32, numpy.int64, numpy.int32, numpy.bool, numpy.bool_],
                          ids=['float64', 'float32', 'int64', 'int32', 'bool', 'bool_'])
 @pytest.mark.parametrize("shape",
-                         [(4,), (2, 3), (2, 2, 2)],
-                         ids=['(4,)', '(2,3)', '(2,2,2)'])
+                         [(0,), (4,), (2, 3), (2, 2, 2)],
```
Comment (samir-nasibli) [inline tests/test_logic.py:45]: `shape = ()` is another case.

**Review by samir-nasibli — APPROVED** [review]

File: tests/test_logic.py (hunk 12 above)
Comment (Rubtsowa) [reply tests/test_logic.py:45]: ia = dpnp.array(a)
dpnp/dpnp_iface_arraycreation.py:205: in array
    return dpnp_array(obj, dtype)
dpnp/dpnp_algo/dpnp_algo.pyx:100: in dpnp.dpnp_algo.dpnp_algo.dpnp_array
    cpdef dparray dpnp_array(obj, dtype=None):
dpnp/dpnp_algo/dpnp_algo.pyx:123: in dpnp.dpnp_algo.dpnp_algo.dpnp_array
    copy_values_to_dparray(result, obj)
dpnp/dpnp_utils/dpnp_algo_utils.pyx:142: in dpnp.dpnp_utils.dpnp_algo_utils.copy_values_to_dparray
    for elem_value in input_obj:
E   TypeError: iteration over a 0-d array

Comment (samir-nasibli) [reply tests/test_logic.py:45]: In `numpy` it works.

For future if there some functionality that doesn't work at the moment, just add needed tests and you can temporarily disabled them by adding them in skip lists.

**Comment by samir-nasibli** [comment]
Seems that these tests (`test_any`, `test_all`) are bad designed, you need some more simple implementation.
